packages/headless/src/app/renew-access-token-middleware.ts

```diff
@@ -0,0 +1,56 @@
+import {Middleware} from '@reduxjs/toolkit';
+import {debounce} from 'ts-debounce';
+import {updateBasicConfiguration} from '../features/configuration/configuration-actions';
+import {ExpiredTokenError} from '../utils/errors';
+
+export function createRenewAccessTokenMiddleware(
+  renewToken: () => Promise<string>
+): Middleware {
+  let accessTokenRenewalsAttempts = 0;
+  const resetRenewalTriesAfterDelay = debounce(
+    () => (accessTokenRenewalsAttempts = 0),
+    500
+  );
+
+  return (store) => (next) => async (action) => {
+    const isThunk = typeof action === 'function';
+
+    if (!isThunk) {
+      return next(action);
+    }
+
+    const payload = await next(action);
+
+    if (!hasExpiredTokenError(payload)) {
+      return payload;
+    }
+
+    accessTokenRenewalsAttempts++;
+    resetRenewalTriesAfterDelay();
+
+    if (accessTokenRenewalsAttempts > 5) {
+      return payload;
```

Comment:
What's the net result for the developer trying to implement renew function that fails 5+ times ?

Should we add some message that explains what's happening ?

What happens if the passed in configuration option `renewAccessToken` is not a function, or not a function that returns a Promise<string>?

I have a feeling that if it fails in this middleware after trying to renew but failing to do so, it's going to be pretty cryptic.

Reply:
There was no feedback, so I made some adjustments following this comment.

If a function is not configured (or something other than is function is configured), when encountering a `419`, the middleware will log a warning explaining that it cannot renew the token, and ask to configure the `renewAccessToken` function when initializing the engine.

The middleware is able to process functions whether they return a promise or not since it is `await`ing the function.

If the `renewAccessToken` function is configured, and the limit is reached, then a different warning is logged, explaining that headless tried but could not renew the token, and to check the `renewAccessToken` function.